LwM2M: Add Multi-fragment Support

mike-scott opened the pull request:
The existing LwM2M framework expects contiguous buffers and this was the reason for the 384 byte buffer sizes.  This was previously a limitation of the ZoAP API.

The new CoAP API doesn't have this limitation and the LwM2M library has already been migrated to use
that API.  Let's finish the process by replacing any contiguous buffer handling with the correct net_pkt APIs to parse across multiple fragments.

Commit:
net: lwm2m: share BUF_ALLOC_TIMEOUT setting w/ other lwm2m components

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: add multi fragment support to LwM2M library

The existing LwM2M framework expected contiguous buffers and this
was the reason for the 384 byte buffer sizes.  This was previously
a limitation of the ZoAP API. The new CoAP API doesn't have this
limitation and the LwM2M library has already been migrated to use
it.

Let's finish the process by replacing any contiguous buffer handling
with the correct net_pkt APIs to parse across multiple fragments.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: lwm2m: remove large buffer size requirement

Now that the LwM2M library can parse across multiple fragements,
let's remove the larger than normal buffer size setting.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1844,11 +1844,14 @@ int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst,
 	data_len = res->data_len;
 
 	/* setup data_ptr/data_len for OPAQUE when it has none setup */
+/* FIXME: */
```
FYI: these 2 FIXME's will be resolved in a future patch dealing with OPAQUE handling.  The current uses are direct writing of firmware to the 5/0/0 (package) resource and bootstrap process for writing DTLS secret.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2021,91 +2027,6 @@ static int lwm2m_delete_handler(struct lwm2m_engine_obj *obj,
-#define COAP_MARKER 0xFF
-
-/* TODO: Remove in favor of using net_buf pointer to payload */
-u8_t *coap_packet_get_payload_ptr(struct coap_packet *cpkt, u16_t *len,
```
@rveerama1 You had asked in an earlier patch about why coap_packet_get_payload_ptr() was being used instead of coap_packet_get_payload() from the CoAP API.  This is the patchset where I remove this function and use coap_packet_get_payload() instead.

rveerama1 replied:
Fine

mike-scott commented:
NOTE: This patch will be rebased once https://github.com/zephyrproject-rtos/zephyr/pull/4554 has been merged.

mike-scott commented:
Rebased on current master to adjust patches for the s*printf -> snprintk change.

mike-scott commented:
Added conf change to remove large buffer setting to this PR where it belongs.

mike-scott commented:
Changing the default buffer size helps with: https://github.com/zephyrproject-rtos/zephyr/issues/3849

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -277,16 +280,35 @@ do_firmware_transfer_reply_cb(const struct coap_packet *response,
+						     opaque_buf);
+			if (!payload_frag && payload_offset == 0xffff) {
+				lwm2m_firmware_set_update_result(
+						RESULT_OUT_OF_MEM);
```
I think this should be a RESULT_INTEGRITY_FAILED instead of OOM.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -667,18 +666,19 @@ static int sm_send_registration(int index, bool send_obj_support_data,
+		ret = coap_packet_append_payload_marker(&msg->cpkt);
+		if (ret < 0) {
+			return ret;
+			goto cleanup;
```
are we doing cleanup or just return??

Commit:
net: lwm2m: introduce getter/setter for OPAQUE data

Each content formatter should have a way of handling opaque data.

For instance TLV data will individually be able to specify a length
but plain text will take up the rest of the packet.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: consolidate URI_LEN define in firmware pull source

No need for 2 different defines to specify URI lengths in the source
for firmware pull method.  Let's combine them.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: remove unused len var from lwm2m_engine_get()

During conversion from the ZoAP to CoAP APIs the use for this variable
was removed, but the variable itself was left in place.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: expose lwm2m_engine_get_resource() for lwm2m subsys

The lwm2m_engine_get_resource() function needs to be made available to
other portions of the lwm2m subsys in order for firmware resource data
to be used in the future.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: fix OPAQUE handling in LwM2M engine

With the change to support multi-fragement buffers in the LwM2M subsys,
the OPAQUE data type was direct write methods were broken.

Let's fix OPAQUE handling by using the newly introduced getter methods
which can use multiple user callbacks (depending on the size of the
user provided buffer).  Let's also add public methods for users to set
/ get OPAQUE data in resources for future use with DTLS key data.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: remove predefined firmware buffer from firmware_pull

Now that we can access resource data in the lwm2m subsys, let's use
the user provided firmware push buffer (5/0/0) to also store the
firmware pull data.

This way the size of the firmware pull buffer is completely up to the
application.

NOTE: This patch adds a 64 byte firmware buffer to the lwm2m_client
sample for this purpose.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Patch V4 is up.  Diff: https://hastebin.com/tukoveduna.cs

This ended up being a larger patchset diff, because I wanted to fix opaque data correctly.

I tested pushing data to the "package" resource of the firmware object by disabling the remote port check in subsys/net/ip/connnection.c and storing the from_addr in the LwM2M udp handler so that I could write back to a port that wasn't the remote connection.

I appreciate everyone's continued help reviewing these patches as we progress to DTLS support. :)  (That's coming next).

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -277,30 +278,62 @@ do_firmware_transfer_reply_cb(const struct coap_packet *response,
+							     len,
+							     write_buf);
+				/* check for end of packet */
+				if (!payload_frag && payload_offset == 0xffff) {
```
This should be treated as an error instead of the end of the packet I suppose.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1835,6 +1848,66 @@ static int lwm2m_read_handler(struct lwm2m_engine_obj_inst *obj_inst,
+
+		if (res->post_write_cb) {
+			/* ignore return value */
+			res->post_write_cb(obj_inst->obj_inst_id,
```
Ok, I'll add a patch correcting the return values of the firmware callbacks to this patchset and then propagate the errors back here.

Commit:
net: lwm2m: fix all return values from resource callbacks

Previously, post_write and execute callbacks returned 1 when handled
and 0 for error condition.  However, this wasn't detailed enough and
the engine can't propagate any sort of error back to users -- so it
doesn't even check the return values in many cases!

Let's adjust the resource callback functions of all objects and the
lwm2m_client sample to return 0 for success or a valid error code.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: propagate errors from post_write callbacks

Now that objects and samples have their return values fixed, let's
propagate them back up to the user if there's an error.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Patch V5 posted.  Diff: https://hastebin.com/zemovovuje.cs

- Addressed @rbtchc 
- Fixed all resource callback functions to return 0 for success and error code otherwise
- Propagate post_write resource callbacks now that valid error codes are returned

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -277,30 +278,62 @@ do_firmware_transfer_reply_cb(const struct coap_packet *response,
+							     write_buf);
+				/* check for end of packet */
+				if (!payload_frag && payload_offset == 0xffff) {
+					return -EINVAL;
```
need to update status before return -EINVAL;
<pre>
lwm2m_firmware_set_update_result(RESULT_INTEGRITY_FAILED)
</pre>

rbtchc replied:
how about this one?

mike-scott replied:
Woops.  Missed this.  Will repost a new version.

rbtchc commented:
I tested the firmware push and pull.
It will crash when doing the firmware push.
Please check the patch for solving the issue: https://hastebin.com/imojecogiy.cpp

mike-scott commented:
Thanks @rbtchc, I'll re-test PUSH method again and check your recommended changes.

mike-scott commented:
Updated with V6 patchset. Diff: https://hastebin.com/ogifomitih.cpp (essentially the patch @rbtchc put in his comments above).

This was a good catch by @rbtchc .  I'm not sure why I don't see the crash (I re-tested both ways and it still seemed to work) -- I'm using QEMU for testing atm as I have some other patches conflicting in a branch where I can test actual hardware.  So that might be the difference.

rbtchc commented:
@mike-scott There was a comment I made yesterday to suggest at _subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c_ line 316.
Here I think we have to change the firmware update result to RESULT_INTEGRITY_FAILED.

mike-scott commented:
Added the update result change mentioned by @rbtchc  Diff: https://hastebin.com/lanamewese.lua
